README
Inspired by the comments at: "http://dk.php.net/time":http://dk.php.net/time and "http://api.rubyonrails.org/classes/ActionView/Helpers/DateHelper.html#M001695":http://api.rubyonrails.org/classes/ActionView/Helpers/DateHelper.html#M001695

h1. Using this scheme to determine the "ago" time:

<pre><code>
  0 <-> 29 secs                                                             # => less than a minute
  30 secs <-> 1 min, 29 secs                                                # => 1 minute
  1 min, 30 secs <-> 44 mins, 29 secs                                       # => [2..44] minutes
  44 mins, 30 secs <-> 89 mins, 29 secs                                     # => about 1 hour
  89 mins, 29 secs <-> 23 hrs, 59 mins, 29 secs                             # => about [2..24] hours
  23 hrs, 59 mins, 29 secs <-> 47 hrs, 59 mins, 29 secs                     # => 1 day
  47 hrs, 59 mins, 29 secs <-> 29 days, 23 hrs, 59 mins, 29 secs            # => [2..29] days
  29 days, 23 hrs, 59 mins, 30 secs <-> 59 days, 23 hrs, 59 mins, 29 secs   # => about 1 month
  59 days, 23 hrs, 59 mins, 30 secs <-> 1 yr minus 1 sec                    # => [2..12] months
  1 yr <-> 2 yrs minus 1 secs                                               # => about 1 year
  2 yrs <-> max time or date                                                # => over [2..X] years
</code></pre>

h1. Using the class

In order to use the class, simply replace your DateTime with WWDateTime.

Before:
<pre><code>
$datetime = new DateTime();
echo $datetime->format(DATE_RFC3339);
</code></pre>

After:
<pre><code>
$datetime = new WWDateTime();
echo $datetime->format(DATE_RFC3339);
</code></pre>

As you can see, there really not that much difference.

In order to get the "Time ago in words" output, all you need to do, is call the 
function @WWDateTime::timeAgoInWords()@

Using the code from before, we get:
<pre><code>
$datetime = new WWDateTime();
echo $datetime->format(DATE_RFC3339);
echo $datetime->timeAgoInWords(); // prints out "less than a minute"
</code></pre>

To test the class properly, we are setting the time 31 seconds back:

<pre><code>
$datetime = new WWDateTime("-31 second");
echo $datetime->format(DATE_RFC3339);
echo $datetime->timeAgoInWords();  // prints out "1 minute"
</code></pre>

h1. Known bugs

* Around midnight the script fails.
* Does not follow my definition completely yet.


Changelog.md
## Changelog

### 3.2.3
* Added a "short" english translation, thanks to @MoPaMo

### 3.2.2
* Update to TR translation to remove three dots from "years" translation, thanks @ozgurg 

### 3.2.1
* Unittest update around leap years

### 3.2.0
* Added Italian translation, thanks to @vegagame
* Added Norweigan translation, thanks to @zreedeecom

### 3.1.1
* Updated a classname psr-4 bug in Zh_Tw, since the file was named Zh_TW.
* Updated php requirement to >=7.0.0, since we are using return types anyways.
* Updated return type in phpdoc on inWordsFromStrings()
* Removed En import in TimeAgo as well.

### 3.1.0
* Changed from psr-0 to psr-4, since some language classes could not be loaded properly
* Added support for DateTimeImmutable objects, thanks to @xtreamwayz
* Added Indonesian translation, thanks to @dhutapratama

### 3.0.6
* Fixed #79 where 1h30m would be 1 hours ago instead of 2 hours ago

### 3.0.5
* Cs translation more correct, thanks @MartkCz

### 3.0.4
* Fixed a bug with isLessThan44Min29Secs, thanks @ruben-haegeman
* Added Polish translations, thanks to @biegacz1

### 3.0.3
* Updated Pt-BR, thanks to @ivesbrito

### 3.0.2
* updated phpdoc for constructor

### 3.0.1
* Typo fixes, thanks @jlratwil
* Did some changes to composer, so php 7.0 can also be tested on travis.
* Removed hhvm and 5.6 from travis, since phpunit cannot go that low.

### 3.0.0
* Redid most of the code, to work with DateTime objects instead.
* Renamed the old inWords method to inWordsFromStrings, so that still works.
* Added a new inWords method to accept DateTime instances.
* Translations are now classes, inheriting from the Language class.
* Fixed the inWords unit tests, to use inWordsFromStrings.
* Added a few unit tests for the new inWords, since both inWords and inWordsFromStrings use the same internals.
* Added a test for changing language.
* Moved changelog things to this file (changelog.md) instead.
* Updated the Readme, to show how to work with the module.

### 2.0.5
* Added Bulgarian translation, thanks to @ironsmile
* Added Latvian translation, thanks to @intarstudents
* Updated French translation, thanks to a team of two? @ArthurHoaro and @blaugueux
* Updated Portuguese translations, thanks to @ivesbrito

### 2.0.4
* Added Catalan translation, thanks to @Ajenbo and @esrefesen

### 2.0.3
* Added Swedish translation, thanks to @khromov
* Added Gregorian translation, thanks to @nikopeikrishvili
* Added Vietnamese translation, thanks to @baivong

### 2.0.2

* Removed user specific files from .gitignore
* Using strict comparison
* Added a method for rounding the number of months to use

### 2.0.1

* Fixed a file encoding problem on the korean translation
* Added a never translate on the turkish translation (thanks to @esrefesen)

### 2.0.0

Now using namespaces to handle the source.

* Added Westsworld namespace to the TimeAgo project.
* Removed the function "timeAgoInWords", since i really didn't like having it around.
* Removed the very deprecated WWDateTime class.
* Fixed unittests after namespace introduction.
* Updated composer.lock
* Updated .travis.yml

### 1.0.0

Introducing exceptions (so remember to handle them) and removed the old way of overloading the translations
path. Also renamed a few methods.
Both things breaks the current API, so I've bumped it to version 1.0.0

* An exception is now thrown, if a translation file cannot be added.
* An exception is now thrown, if the translation file is empty.
* Made timechecks a lot easier to read, by moving them into a "isLessThanX" method.
* Added unittests to test the loading of translations.
* Added unittest for dateDifference.
* Removed: _loadTranslations($language, $alternativePath), you can no longer use alternativePath, if extending the
component. Instead I added a define called TIMEAGO_TRANSLATION_PATH, which can be used to override the translations path
* Renamed _loadTranslations to loadTranslations
* Renamed _translate to translate

### 0.4.15
Added Turkish translation (@esrefesen), and Thai (@jaideejung007) and iranian- Farsi(Persian) (@datamweb) and French (@lamberttraccard)

### 0.4.14
Added Arabic translation (@xc0d3rz)
### 0.4.13
@jmontoyaa fixed a timezone setting bug
### 0.4.12
Merged #18, #33 (you can now supply an alternative translations folder), #35 and Added Czech (@misaon)
### 0.4.11
Changes to French translation (@souhailmerroun). Added Russian and Ukranian translations (@akavolkol)
### 0.4.10
Added Korean and Finnish translations (thanks to @easylogic and @tk1sul1)
### 0.4.9
Added Spanish and Hungarian and Brazilian Portuguese (thanks to @khrizt and @technodelight and @gugoan)
### 0.4.8
Added French and Japanese and Taiwanese translations (thanks to @barohatoum and @hotengchang)
### 0.4.7
Changes to fix annoying "floor" bugs.
### 0.4.6
Added Dutch and Chinese translations (thanks to @NielsdeBlaauw and @su-xiaolin)
### 0.4.5
Fix for wrongly named variable - patch from @rimager
### 0.4.4
Changed the previous fix, since it caused a problem with the "Scheme" at the start of this document. "about 1 hour ago" is between 44mins30secs and 89mins29secs.
### 0.4.3
Added a fix for "about 1 hour ago", which would yield "1 hours ago" if the time was 1hour 30minutes. Giving a pluralization error.
### 0.4.2
Added German translations .
### 0.4.1
Added "ago" to the end of the translations, so the texts now read "about 1 hour ago".
### 0.4.0
Added translations (Danish and English) to the plugin.

# The old way of doing things.

Just if you wanted to try out the old "time ago" class, I've kept it in the project.
It will eventually be deleted, but hey, go have fun with it. Perhaps even extend it etc.
Bear in mind though, that the new DateTime::diff in PHP version 5.3.0 will eventually
replace it completely. (PHP 5.3.0 DateTime::diff link: http://www.php.net/manual/en/datetime.diff.php)


timeago.inc.php
<?php

// To allow overriding this function as will, if someone wants to use a derived class of `TimeAgo`
if (!function_exists('timeAgoInWords')) {
    function timeAgoInWords($timestring, $timezone = null, $language = 'en')
    {
        $timeAgo = new TimeAgo($timezone, $language);

        return $timeAgo->inWords($timestring, "now");
    }
}


/**
 * This class can help you find out just how much time has passed between
 * two dates.
 *
 * It has two functions you can call:
 * inWords() which gives you the "time ago in words" between two dates.
 * dateDifference() which returns an array of years,months,days,hours,minutes and
 * seconds between the two dates.
 *
 * @author jimmiw
 * @since 0.2.0 (2010/05/05)
 * @site http://github.com/jimmiw/php-time-ago
 */
class TimeAgo
{
    // defines the number of seconds per "unit"
    private $secondsPerMinute = 60;
    private $secondsPerHour = 3600;
    private $secondsPerDay = 86400;
    private $secondsPerMonth = 2592000;
    private $secondsPerYear = 31536000; // 31622400 seconds on leap years though...
    private $timezone;
    private $previousTimezone;

    // translations variables
    private static $language;
    private static $timeAgoStrings = null;

    /**
     * TimeAgo constructor.
     * @param null|DateTimeZone $timezone the timezone to use (uses system if none is given)
     * @param string $language the language to use (defaults to 'en' for english)
     */
    public function __construct($timezone = null, $language = 'en')
    {
        // loads the translation files
        self::loadTranslations($language);
        // storing the current timezone
        $this->timezone = $timezone;
    }

    /**
     * Fetches the different between $past and $now in a spoken format.
     * NOTE: both past and now should be parseable by strtotime
     * @param string $past the past date to use
     * @param string $now the current time, defaults to now (can be an other time though)
     * @return string the difference in spoken format, e.g. 1 day ago
     */
    public function inWords($past, $now = "now")
    {
        // sets the default timezone
        $this->changeTimezone();
        // finds the past in datetime
        $past = strtotime($past);
        // finds the current datetime
        $now = strtotime($now);

        // creates the "time ago" string. This always starts with an "about..."
        $timeAgo = "";

        // finds the time difference
        $timeDifference = $now - $past;

        // rule 0
        // $past is null or empty or ''
        if ($this->isPastEmpty($past)) {
            $timeAgo = $this->translate('never');
        }
        // rule 1
        // less than 29secs
        else if ($this->isLessThan29Seconds($timeDifference)) {
            $timeAgo = $this->translate('lessThanAMinute');
        }
        // rule 2
        // more than 29secs and less than 1min29secss
        else if ($this->isLessThan1Min29Seconds($timeDifference)) {
            $timeAgo = $this->translate('oneMinute');
        }
        // rule 3
        // between 1min30secs and 44mins29secs
        else if ($this->isLessThan44Min29Secs($timeDifference)) {
            $minutes = round($timeDifference / $this->secondsPerMinute);
            $timeAgo = $this->translate('lessThanOneHour', $minutes);
        }
        // rule 4
        // between 44mins30secs and 1hour29mins59secs
        else if ($this->isLessThan1Hour29Mins59Seconds($timeDifference)) {
            $timeAgo = $this->translate('aboutOneHour');
        }
        // rule 5
        // between 1hour29mins59secs and 23hours59mins29secs
        else if ($this->isLessThan23Hours59Mins29Seconds($timeDifference)) {
            $hours = round($timeDifference / $this->secondsPerHour);
            $timeAgo = $this->translate('hours', $hours);
        }
        // rule 6
        // between 23hours59mins30secs and 47hours59mins29secs
        else if ($this->isLessThan47Hours59Mins29Seconds($timeDifference)) {
            $timeAgo = $this->translate('aboutOneDay');
        }
        // rule 7
        // between 47hours59mins30secs and 29days23hours59mins29secs
        else if ($this->isLessThan29Days23Hours59Mins29Seconds($timeDifference)) {
            $days = round($timeDifference / $this->secondsPerDay);
            $timeAgo = $this->translate('days', $days);
        }
        // rule 8
        // between 29days23hours59mins30secs and 59days23hours59mins29secs
        else if ($this->isLessThan59Days23Hours59Mins29Secs($timeDifference)) {
            $timeAgo = $this->translate('aboutOneMonth');
        }
        // rule 9
        // between 59days23hours59mins30secs and 1year (minus 1sec)
        else if ($this->isLessThan1Year($timeDifference)) {
            $months = round($timeDifference / $this->secondsPerMonth);
            // if months is 1, then set it to 2, because we are "past" 1 month
            if ($months == 1) {
                $months = 2;
            }

            $timeAgo = $this->translate('months', $months);
        }
        // rule 10
        // between 1year and 2years (minus 1sec)
        else if ($this->isLessThan2Years($timeDifference)) {
            $timeAgo = $this->translate('aboutOneYear');
        }
        // rule 11
        // 2years or more
        else {
            $years = floor($timeDifference / $this->secondsPerYear);
            $timeAgo = $this->translate('years', $years);
        }

        $this->restoreTimezone();

        return $timeAgo;
    }

    /**
     * Fetches the date difference between the two given dates.
     * NOTE: both past and now should be parseable by strtotime
     *
     * @param string $past the "past" time to parse
     * @param string $now the "now" time to parse
     * @return array the difference in dates, using the two dates
     */
    public function dateDifference($past, $now = "now")
    {
        // initializes the placeholders for the different "times"
        $seconds = 0;
        $minutes = 0;
        $hours = 0;
        $days = 0;
        $months = 0;
        $years = 0;

        // sets the default timezone
        $this->changeTimezone();

        // finds the past in datetime
        $past = strtotime($past);
        // finds the current datetime
        $now = strtotime($now);

        // calculates the difference
        $timeDifference = $now - $past;

        // starts determining the time difference
        if ($timeDifference >= 0) {
            switch ($timeDifference) {
                // finds the number of years
                case ($timeDifference >= $this->secondsPerYear):
                    // uses floor to remove decimals
                    $years = floor($timeDifference / $this->secondsPerYear);
                    // saves the amount of seconds left
                    $timeDifference = $timeDifference - ($years * $this->secondsPerYear);

                // finds the number of months
                case ($timeDifference >= $this->secondsPerMonth && $timeDifference <= ($this->secondsPerYear - 1)):
                    // uses floor to remove decimals
                    $months = floor($timeDifference / $this->secondsPerMonth);
                    // saves the amount of seconds left
                    $timeDifference = $timeDifference - ($months * $this->secondsPerMonth);

                // finds the number of days
                case ($timeDifference >= $this->secondsPerDay && $timeDifference <= ($this->secondsPerYear - 1)):
                    // uses floor to remove decimals
                    $days = floor($timeDifference / $this->secondsPerDay);
                    // saves the amount of seconds left
                    $timeDifference = $timeDifference - ($days * $this->secondsPerDay);

                // finds the number of hours
                case ($timeDifference >= $this->secondsPerHour && $timeDifference <= ($this->secondsPerDay - 1)):
                    // uses floor to remove decimals
                    $hours = floor($timeDifference / $this->secondsPerHour);
                    // saves the amount of seconds left
                    $timeDifference = $timeDifference - ($hours * $this->secondsPerHour);

                // finds the number of minutes
                case ($timeDifference >= $this->secondsPerMinute && $timeDifference <= ($this->secondsPerHour - 1)):
                    // uses floor to remove decimals
                    $minutes = floor($timeDifference / $this->secondsPerMinute);
                    // saves the amount of seconds left
                    $timeDifference = $timeDifference - ($minutes * $this->secondsPerMinute);

                // finds the number of seconds
                case ($timeDifference <= ($this->secondsPerMinute - 1)):
                    // seconds is just what there is in the timeDifference variable
                    $seconds = $timeDifference;
            }
        }

        $this->restoreTimezone();

        $difference = [
            "years" => $years,
            "months" => $months,
            "days" => $days,
            "hours" => $hours,
            "minutes" => $minutes,
            "seconds" => $seconds,
        ];

        return $difference;
    }

    /**
     * Translates the given $label, and adds the given $time.
     * @param string $label the label to translate
     * @param string $time the time to add to the translated text.
     * @return string the translated label text including the time.
     */
    protected function translate($label, $time = '')
    {
        // handles a usecase introduced in #18, where a new translation was added.
        // This would cause an array-out-of-bound exception, since the index does not
        // exist in most translations.
        if (!isset(self::$timeAgoStrings[$label])) {
            return '';
        }

        return sprintf(self::$timeAgoStrings[$label], $time);
    }

    /**
     * Loads the translations into the system.
     * NOTE: Removed alternativePath in 0.6.0, instead, define that path using TIMEAGO_TRANSLATION_PATH
     * @param string $language the language iso to use
     * @throws Exception if a language file cannot be found or there are no translations
     */
    protected static function loadTranslations($language)
    {
        // no time strings loaded? load them and store it all in static variables
        if (self::$timeAgoStrings == null || self::$language != $language) {
            // default path to the translations
            $basePath = __DIR__ . '/translations/';

            // adding the possibility for an alternate translations path
            if (defined('TIMEAGO_TRANSLATION_PATH')) {
                $basePath = TIMEAGO_TRANSLATION_PATH;
            }

            // setting the translation path
            $path = $basePath . $language . '.php';

            if (! file_exists($path)) {
                throw new Exception("No translation file found at: " . $path);
            }

            // loads the translation file
            include($path);

            // throws an exception, if the are no translations, in the currently loaded translation file
            if (! isset($timeAgoStrings)) {
                throw new Exception('No translations found in translation file at: ' . $path);
            }

            // storing the time strings in the current object
            self::$timeAgoStrings = $timeAgoStrings;
        }

        // storing the language
        self::$language = $language;
    }

    /**
     * Changes the timezone
     */
    protected function changeTimezone()
    {
        $this->previousTimezone = false;
        if ($this->timezone) {
            $this->previousTimezone = date_default_timezone_get();
            date_default_timezone_set($this->timezone);
        }
    }

    /**
     * Restores a previous timezone
     */
    protected function restoreTimezone()
    {
        if ($this->previousTimezone) {
            date_default_timezone_set($this->previousTimezone);
            $this->previousTimezone = false;
        }
    }

    /**
     * Checks if the given past is empty
     * @param string $past the "past" to check
     * @return bool true if empty, else false
     */
    private function isPastEmpty($past)
    {
        return $past === '' || is_null($past) || empty($past);
    }

    /**
     * Checks if the time difference is less than 29seconds
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan29Seconds($timeDifference)
    {
        return $timeDifference <= 29;
    }

    /**
     * Checks if the time difference is less than 1min 29seconds
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan1Min29Seconds($timeDifference)
    {
        return $timeDifference >= 30 && $timeDifference <= 89;
    }

    /**
     * Checks if the time difference is less than 44mins 29seconds
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan44Min29Secs($timeDifference)
    {
        return $timeDifference >= 90 &&
        $timeDifference <= (($this->secondsPerMinute * 44) + 29);
    }

    /**
     * Checks if the time difference is less than 1hour 29mins 59seconds
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan1Hour29Mins59Seconds($timeDifference)
    {
        return $timeDifference >= (($this->secondsPerMinute * 44) + 30)
        &&
        $timeDifference <= ($this->secondsPerHour + ($this->secondsPerMinute * 29) + 59);
    }

    /**
     * Checks if the time difference is less than 23hours 59mins 29seconds
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan23Hours59Mins29Seconds($timeDifference)
    {
        return $timeDifference >= (
            $this->secondsPerHour +
            ($this->secondsPerMinute * 30)
        )
        &&
        $timeDifference <= (
            ($this->secondsPerHour * 23) +
            ($this->secondsPerMinute * 59) +
            29
        );
    }

    /**
     * Checks if the time difference is less than 27hours 59mins 29seconds
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan47Hours59Mins29Seconds($timeDifference)
    {
        return $timeDifference >= (
            ($this->secondsPerHour * 23) +
            ($this->secondsPerMinute * 59) +
            30
        )
        &&
        $timeDifference <= (
            ($this->secondsPerHour * 47) +
            ($this->secondsPerMinute * 59) +
            29
        );
    }

    /**
     * Checks if the time difference is less than 29days 23hours 59mins 29seconds
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan29Days23Hours59Mins29Seconds($timeDifference)
    {
        return $timeDifference >= (
            ($this->secondsPerHour * 47) +
            ($this->secondsPerMinute * 59) +
            30
        )
        &&
        $timeDifference <= (
            ($this->secondsPerDay * 29) +
            ($this->secondsPerHour * 23) +
            ($this->secondsPerMinute * 59) +
            29
        );
    }

    /**
     * Checks if the time difference is less than 59days 23hours 59mins 29seconds
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan59Days23Hours59Mins29Secs($timeDifference)
    {
        return $timeDifference >= (
            ($this->secondsPerDay * 29) +
            ($this->secondsPerHour * 23) +
            ($this->secondsPerMinute * 59) +
            30
        )
        &&
        $timeDifference <= (
            ($this->secondsPerDay * 59) +
            ($this->secondsPerHour * 23) +
            ($this->secondsPerMinute * 59) +
            29
        );
    }

    /**
     * Checks if the time difference is less than 1 year
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan1Year($timeDifference)
    {
        return $timeDifference >= (
            ($this->secondsPerDay * 59) +
            ($this->secondsPerHour * 23) +
            ($this->secondsPerMinute * 59) +
            30
        )
        &&
        $timeDifference < $this->secondsPerYear;
    }

    /**
     * Checks if the time difference is less than 2 years
     * @param int $timeDifference the time difference in seconds
     * @return bool
     */
    private function isLessThan2Years($timeDifference)
    {
        return $timeDifference >= $this->secondsPerYear
        &&
        $timeDifference < ($this->secondsPerYear * 2);
    }
}
